pydicom/encoders/gdcm.py

```diff
@@ -0,0 +1,221 @@
+        )
+
+    return _ENCODERS[kwargs['transfer_syntax_uid']](src, **kwargs)
+
+
+def _rle_encode(src, **kwargs):
+    """Return RLE encoded image data from `src`.
+
+    Parameters
+    ----------
+    src : bytes
+        The raw image frame data to be encoded.
+    **kwargs
+        Required parameters:
+
+        * `rows`: int
+        * `columns`: int
+        * `samples_per_pixel`: int
+        * `number_of_frames`: int
+        * `bits_allocated`: int
+        * `bits_stored`: int
+        * `pixel_representation`: int
+        * `photometric_interpretation`: str
+
+    Returns
+    -------
+    bytes
+        The encoded image data.
+    """
+    # Check the parameters are valid for RLE encoding with GDCM
+    samples_per_pixel: int = kwargs['samples_per_pixel']
+    bits_allocated: int = kwargs['bits_allocated']
+
+    # Bug up to v3.0.9 (Apr 2021) in handling 32-bit, 3 sample/px data
+    gdcm_version = [int(c) for c in gdcm.Version.GetVersion().split('.')]
+    if gdcm_version < [3, 1, 0]:
+        if bits_allocated == 32 and samples_per_pixel == 3:
+            raise RuntimeError(
+                "The 'gdcm' plugin is unable to RLE encode 32-bit, 3 "
+                "samples/px data with GDCM v3.0.9 or older"
+            )
+
+    if bits_allocated > 32:
+        raise ValueError(
+            f"The 'gdcm' plugin is unable to encode {bits_allocated}-bit data"
+        )
+
+    # Create a gdcm.Image with the uncompressed `src` data
+    image = _create_gdcm_image(src, **kwargs)
+
+    # Converts an image to match the set transfer syntax
+    converter = gdcm.ImageChangeTransferSyntax()
+
+    # Set up the converter with the intended transfer syntax...
+    ts = gdcm.TransferSyntax.GetTSType(kwargs['transfer_syntax_uid'])
+    converter.SetTransferSyntax(gdcm.TransferSyntax(ts))
+    # ...and image to be converted
+    converter.SetInput(image)
+
+    # Perform the conversion, returns bool
+    # 'PALETTE COLOR' and a lossy transfer syntax will return False
+    result = converter.Change()
+    if not result:
+        raise RuntimeError(
+            "An error occurred with the 'gdcm' plugin: "
+            "ImageChangeTransferSyntax.Change() returned a failure result"
+        )
+
+    # The converted image as gdcm.Image
+    # Weirdly, reusing `image` as the variable name causes a segfault here
+    output = converter.GetOutput()
+    # The element's value is the encapsulated encoded pixel data
+    seq = output.GetDataElement().GetSequenceOfFragments()
+
+    # RLECodec::Code() uses only 1 fragment per frame
+    if seq is None or seq.GetNumberOfFragments() != 1:
+        # Covers both no sequence and unexpected number of fragments
+        raise RuntimeError(
+            "An error occurred with the 'gdcm' plugin: unexpected number of "
+            "fragments found in the 'Pixel Data'"
+        )
+
+    fragment = seq.GetFragment(0)
+    return (
+        fragment.GetByteValue().GetBuffer().encode("utf-8", "surrogateescape")
```

Comment:
Why utf-8-encoded? I don't know GDCM, so I'm just ignorant here...

Reply:
GDCM returns a string here rather than bytes. I reused the same technique from the GDCM decoder but I think it originally came from [this mailing list issue](https://sourceforge.net/p/gdcm/mailman/gdcm-developers/thread/CA%2B7wUszUuPQUG-wquZR9WH0O4L7oHAacXRX2m8nhFaaRnuienQ%40mail.gmail.com/).

Hmm, I wonder if the python-gdcm project would accept that patch Mathieu mentions.

Reply:
The [linked SWIG docs](http://www.swig.org/Doc3.0/Python.html#Python_nn77) explain why.

Reply:
I see - interesting, though this is certainly not the most efficient way to do this.